File: lang/c/src/vpkppc/format/VPK.cpp
#include <vpkppc/format/VPK.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const uint32_t VPKPP_FPX_SIGNATURE = FPX_SIGNATURE;
const char* VPKPP_FPX_DIR_SUFFIX   = FPX_DIR_SUFFIX.data();
const char* VPKPP_FPX_EXTENSION    = FPX_EXTENSION.data();

const uint32_t VPKPP_VPK_SIGNATURE = VPK_SIGNATURE;
const char* VPKPP_VPK_DIR_SUFFIX   = VPK_DIR_SUFFIX.data();
const char* VPKPP_VPK_EXTENSION    = VPK_EXTENSION.data();

const uint16_t VPKPP_VPK_DIR_ARCHIVE_INDEX = VPK_DIR_ARCHIVE_INDEX;

const uint16_t VPKPP_VPK_MAX_PRELOAD_BYTES = VPK_MAX_PRELOAD_BYTES;

const uint32_t VPKPP_VPK_DEFAULT_CHUNK_SIZE = VPK_DEFAULT_CHUNK_SIZE;

VPKPP_API vpkpp_pack_file_handle_t vpkpp_vpk_create(const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = VPK::create(path);
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

VPKPP_API vpkpp_pack_file_handle_t vpkpp_vpk_create_with_options(const char* path, vpkpp_vpk_version_e version) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = VPK::create(path, convert::cast(version));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

VPKPP_API vpkpp_pack_file_handle_t vpkpp_vpk_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = VPK::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

VPKPP_API void vpkpp_vpk_generate_keypair_files(const char* base, vpkpp_vpk_signature_type_e signatureType) {
	SOURCEPP_EARLY_RETURN(base);

	return VPK::generateKeyPairFiles(base, convert::cast(signatureType));
}

VPKPP_API int vpkpp_vpk_sign_from_file(vpkpp_pack_file_handle_t handle, const char* filename, vpkpp_vpk_signature_type_e signatureType) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(filename, false);

	auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(vpk, false);

	return vpk->sign(filename, convert::cast(signatureType));
}

VPKPP_API int vpkpp_vpk_sign_from_mem(vpkpp_pack_file_handle_t handle, const unsigned char* privateKeyBuffer, size_t privateKeyLen, const unsigned char* publicKeyBuffer, size_t publicKeyLen, vpkpp_vpk_signature_type_e signatureType) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(privateKeyBuffer, false);
	SOURCEPP_EARLY_RETURN_VAL(privateKeyLen, false);
	SOURCEPP_EARLY_RETURN_VAL(publicKeyBuffer, false);
	SOURCEPP_EARLY_RETURN_VAL(publicKeyLen, false);

	auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(vpk, false);

	return vpk->sign(
		{reinterpret_cast<const std::byte*>(privateKeyBuffer), reinterpret_cast<const std::byte*>(privateKeyBuffer + privateKeyLen)},
		{reinterpret_cast<const std::byte*>(publicKeyBuffer), reinterpret_cast<const std::byte*>(publicKeyBuffer + publicKeyLen)},
		convert::cast(signatureType));
}

VPKPP_API vpkpp_vpk_version_e vpkpp_vpk_get_version(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, VPKPP_VPK_VERSION_VALVE_V0);

	const auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(vpk, VPKPP_VPK_VERSION_VALVE_V0);

	return convert::cast(vpk->getVersion());
}

VPKPP_API void vpkpp_vpk_set_version(vpkpp_pack_file_handle_t handle, vpkpp_vpk_version_e version) {
	SOURCEPP_EARLY_RETURN(handle);

	auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN(vpk);

	vpk->setVersion(convert::cast(version));
}

VPKPP_API uint32_t vpkpp_vpk_get_chunk_size(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	const auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(vpk, 0);

	return vpk->getChunkSize();
}

VPKPP_API void vpkpp_vpk_set_chunk_size(vpkpp_pack_file_handle_t handle, uint32_t chunkSize) {
	SOURCEPP_EARLY_RETURN(handle);

	auto* vpk = dynamic_cast<VPK*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN(vpk);

	vpk->setChunkSize(chunkSize);
}

File: lang/csharp/src/vpkpp/Format/VPK.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class VPK : PackFile
{
	protected VPK(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public static VPK? Create(string path)
	{
		var handle = DLL.vpkpp_vpk_create(path);
		return handle == nint.Zero ? null : new VPK(handle);
	}

	public static VPK? Create(string path, uint version)
	{
		var handle = DLL.vpkpp_vpk_create_with_options(path, version);
		return handle == nint.Zero ? null : new VPK(handle);
	}

	public new static VPK? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_vpk_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new VPK(handle);
	}

	public static bool GenerateKeypairFiles(string path)
	{
		return Convert.ToBoolean(DLL.vpkpp_vpk_generate_keypair_files(path));
	}

	public bool Sign(string filename)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vpkpp_vpk_sign_from_file(Handle, filename));
	}

	public bool Sign(ReadOnlySpan<byte> privateKeyBuffer, ReadOnlySpan<byte> publicKeyBuffer)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vpkpp_vpk_sign_from_mem(Handle, privateKeyBuffer, (ulong) privateKeyBuffer.Length, publicKeyBuffer, (ulong) publicKeyBuffer.Length));
	}

	public uint Version
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vpkpp_vpk_get_version(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vpkpp_vpk_set_version(Handle, value);
		}
	}
	
	public uint ChunkSize
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vpkpp_vpk_get_chunk_size(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vpkpp_vpk_set_chunk_size(Handle, value);
		}
	}
}
